### ISSUE TITLE
See if we can run LFE in restricted mode

### ISSUE THREAD

**Issue by oubiwann** [lfex/try.lfe.io#3]

**Comment by yurrriq** [comment]
I'm experimenting with this now.

**Edit**: see results in [this gist](https://gist.github.com/yurrriq/d3d1a3d435685b13f2fa)

**Comment by yurrriq** [comment]
Looks like there would need to be some changes made to [lfe/src/lfe_shell.erl](https://github.com/rvirding/lfe/blob/develop/src/lfe_shell.erl) to store and check the restricted status. Unless I'm mistaken, it seems reasonably straightforward after reading the pertinent parts of [otp/lib/stdlib/src/shell.erl](https://github.com/erlang/otp/blob/master/lib/stdlib/src/shell.erl).

I guess I would prefer a "proper" solution like implementing restricted LFE shells, rather than something more ad-hoc, but I'm not opposed to hacking together something that works instead, if needed.

**Comment by yurrriq** [comment]
[More notes](http://erlang.org/pipermail/erlang-bugs/2012-January/002742.html) from the [Try Erlang](http://www.tryerlang.org/) camp.

**Comment by oubiwann** [comment]
Good stuff -- thanks for the notes!

**Comment by rvirding** [comment]
Yes, I am very sceptical of trying to make a good restricted shell, with enough levels of meta-programming you are always going to get around it. And once you start running one of the standard libraries you totally lose control. Didn't the tryerlang go with some form of sandboxing?

**Comment by yurrriq** [comment]
Hmm yeah... I can't remember how they did it at the moment and despite the promise, I couldn't find the code on GH. I wrote a Clojure RPC for work once that used a pretty basic whitelist and checked all the funcall spots. After spotting a few weird vulnerabilities it ended up working pretty well (and securely). Maybe that's a better approach.

We could probably repurpose some of the [Pynchon macros](https://github.com/yurrriq/pynchon/blob/master/include/arrows.lfe) to preprocess incoming forms. That, in concert with some jailing, would probably work out nicely.

I've also just discovered [GoTTY](https://github.com/yudai/gotty), which may prove helpful.
